src/lib/comms/sol-http-client-impl-curl.c

```diff
@@ -685,6 +685,53 @@ set_post_fields_from_params(CURL *curl, struct sol_arena *arena,
 }
 
 static bool
+set_post_data_from_params(CURL *curl, struct sol_arena *arena,
+    const struct sol_http_param *params)
+{
+    struct sol_buffer *data = NULL;
+    struct sol_http_param_value *iter;
+    uint16_t idx;
+    char *tmp;
+    bool type_set = false;
+
+    SOL_VECTOR_FOREACH_IDX (&params->params, iter, idx) {
+        const char *key = iter->value.key_value.key;
+        struct sol_buffer *value = iter->value.data.value;
+
+        if (iter->type == SOL_HTTP_PARAM_POST_FIELD) {
+            SOL_WRN("SOL_HTTP_PARAM_POST_FIELD and SOL_HTTP_PARAM_POST_DATA found in parameters."
+                   " Only one can be used at a time");
+            return false;
+        } else if (iter->type == SOL_HTTP_PARAM_HEADER) {
+            type_set = type_set || strcmp(key, "content-type");
+        } else  if (iter->type == SOL_HTTP_PARAM_POST_DATA) {
+            if (data) {
+                SOL_WRN("More than one SOL_HTTP_PARAM_POST_DATA found.");
+                return false;
+            }
+
+            data = value;
+        }
+    }
+
+    if (!data)
+        return false;
+
+    if (!type_set)
+        SOL_WRN("POST request has data but no content-type was set");
+
+    tmp = strndup((char *)data->data, (int)data->used);
+    if (!tmp)
+        goto cleanup;
+
+    return set_string_option(curl, CURLOPT_POSTFIELDS, arena, tmp);
+
+cleanup:
```

Comment:
No need to have this label, just `return false` on line 725. (Or just use `SOL_NULL_CHECK()`.)